File: arrays/smallest-difference.js
// time: O(n(log(n)) + m(log(m)))
// space: O(1)
// where n is the first array and m is the second
function smallestDifference(arrayOne, arrayTwo) {
	arrayOne.sort((a, b) => a - b)
	arrayTwo.sort((a, b) => a - b)
  let idx1 = 0
	let idx2 = 0
	let smallest = Infinity
	let current = Infinity
	let smallestPair = []
	while (idx1 < arrayOne.length && idx2 < arrayTwo.length) {
		let first = arrayOne[idx1]
		let second = arrayTwo[idx2]
		if (first < second) {
			current = second - first
			idx1++
		} else if (second < first) {
			current = first - second
			idx2++
		} else return [first, second]
		if (smallest > current) {
			smallest = current
			smallestPair = [first, second]
		}
	}
	return smallestPair
}

smallestDifference([-1, 5, 10, 20, 28, 3], [26, 134, 135, 15, 17]) // [28, 26]

File: algo-expert/medium/arrays/smallest-difference.py
# time: O(nlog(n) + mlog(m))
# space: O(1)
# where n and m are the lengths of the two input arrays
def smallestDifference(arrayOne, arrayTwo):
    arrayOne.sort()
    arrayTwo.sort()
    i = 0
    j = 0
    smallest = float("inf")
    current = float("inf")
    smallestPair = []
    while i < len(arrayOne) and j < len(arrayTwo):
        first = arrayOne[i]
        second = arrayTwo[j]
        if first < second:
            current = second - first
            i += 1
        elif second < first:
            current = first - second
            j += 1
        else:
            return [first, second]
        if current < smallest:
            smallest = current
            smallestPair = [first, second]
    return smallestPair

print(smallestDifference(one, two))
